python/pyspark/serializers.py

```diff
@@ -65,7 +65,7 @@
     from itertools import izip as zip, imap as map
 else:
     import pickle
-    protocol = 3
+    protocol = min(pickle.HIGHEST_PROTOCOL, 4)
```

Comment:
I actually think we can always use `HIGHEST_PROTOCOL` because PySpark's going to throw an exception for mismatched Python versions anyway. Cloudpickle also does that. Let's match it.

Reply:
Done. Also reorganised older commits a bit.